docs/proposals/event-notification.md

```diff
@@ -0,0 +1,141 @@
+# ODO build and application status notification 
+
+## Summary
+
+With this proposal and [it's related issue](https://github.com/openshift/odo/issues/2550) we examine how to consume build/run event output from odo in order to allow external tools (such as IDEs) to determine the application/build status of an odo-managed application.
+
+In short, the proposal is:
+- New flag for push command, `odo push -o json`: Outputs JSON events that correspond to what devfile actions/commands are being executed by push.
+- New odo command `odo component status -o json`: Pings the application URLs, and checks the container/pod status every X seconds (and/or a watch, for Kubernetes case). The result of both is output as JSON to be used by external tools to determine if the application is running.
```

Comment:
What if application doesn't have any URLs?
Also, what would be considered as pinging url? Just checking if the port is open or sending HTTP commands like GET or HEAD?

Wouldn't it be enough to just check if the application process is running?
That could be done easily, as it will run as a supervisord service. There could be additional information in the status message like if the URL as alive or not (using HTTP HEAD), but the primary source for the status would be if the process  is running or not.

Reply:
Good questions:

If the user is developing a web application, often the first metric of "is my application running?" is "can I see it in my browser when I refresh?". I attempt to capture this idea with this part of the proposal.

The algorithm I had considered would be:
- For each odo-managed url:
  - issue a GET/HEAD request to the app domain root, for example a component 'my-app' might be pinged at https://my-app.my-cluster-domain.com:8080
  - `status` command to output a success to the console if ANY HTTP response (2xx/3xx/4xx/5xx) code is received, since what we are testing is whether the app is connectable (a failure would be no response before timeout).

Now this raises a few good question: 
- Is it always safe (eg [safe](https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol#Safe_methods), non-destructive, immutable, idempotent) to issue GET/HEAD requests to the root of an any odo-managed application without the user's explicit acknowledgement? I had assumed yes (eg low risk), but am definitely open to further discussion here. 
- Will Kubernetes ingress controllers ALWAYS respond with an HTTP response, even if sent to an invalid cluster URL? If so, this would make this check not especially useful for the Kubernetes case...
- Could we just remove this entire algorithm from the proposal and leave it up to the IDE to ping the URL if appropriate? After all, the IDE does have full knowledge over which URLs are exposed via odo... at the expense of slightly more work on the IDE-side.

As for the case where no application URLs are exposed (and there are many such applications): if no odo-managed URLs are defined, the algorithm will be a no-op, and no URL success/failure output will be output.

RE: checking if the application process is running, I like the idea and is definitely preferable if feasible, I will expand on this further [on the other comment where you mentioned this](https://github.com/openshift/odo/pull/3177#discussion_r425430516).